backend/controllers/eventsController.js

```diff
@@ -1,6 +1,122 @@
 const ErrorResponse = require('../utils/errorResponse');
 const asyncHandler = require('../middleware/async');
 const Event = require('../models/Event');
+const Vehicle = require('../models/Vehicle');
+
+// @desc        Create an event for a specific vehicle
+// @access      Private
+exports.createEvent = asyncHandler(async (req, res, next) => {
+  // This endpoint is used for creating new events for users
+
+  // Vehicle APIs
+  // DELETE /api/v1/inventory/vehicle/:vehicleId
+  // PUT /api/v1/inventory/vehicle/:vehicleId
+
+  //Transaction APIs
+  // PUT /api/v1/inventory/details/sale/:saleId'
+  // POST /api/v1/inventory/details/sale'
+
+  // 'vehicle_sale_pending', DONE
+  // 'vehicle_approved', DONE
+  // 'vehicle_delivered', DONE
+  // 'vehicle_missing', DONE
+  // 'vehicle_moved', CAn'T DO FOR NOW
+  // 'vehicle_found', DONE
+  // 'vehicle_deleted', DONE
+  // 'transaction_modified', DONE
+
+  var body = {};
+  //Transaction API Events
+  if (req.originalUrl.indexOf('/api/v1/inventory/details/sale') > -1) {
+    if (req.method=='POST')
+    {
+      //fetch vehicle details
+      const vehicle = await Vehicle.findById(req.body.vehicle);
+      body = {
+        event_type: 'vehicle_sale_pending',
+        dealership: req.body.dealership,
+        vehicle: req.body.vehicle,
+        title: 'New Transaction',
+        description: `A new transaction has been created for the vehicle with vin: ${vehicle.vin} by ${req.user.first_name} ${req.user.last_name}`,
+      }
+    }
+    // When a transaction has been approved. EventType=vehicle_approved
+    else if (req.method=='PUT') {
+      const vehicle = await Vehicle.find({ sale: req.params.saleId });
+      if (!!req.body.date_approved) {
+        body = {
+          event_type: 'vehicle_approved',
+          dealership: vehicle[0].dealership,
+          vehicle: vehicle[0]._id,
+          title: 'Transaction Approved',
+          description: `A sale request has been approved for the vehicle with vin: ${vehicle[0].vin} by ${req.user.first_name} ${req.user.last_name}`,
+        }
+      }
+      else if (!!req.body.sale_amount) {
+        body = {
+          event_type: 'transaction_modified',
+          dealership: vehicle[0].dealership,
+          vehicle: vehicle[0]._id,
+          title: 'Transaction Modified',
+          description: `A transaction has been modified for the vehicle with vin: ${vehicle[0].vin} by ${req.user.first_name} ${req.user.last_name}`,
+        }
+      }
+    }
+  }
+
+  //Vehicle API Events
+  else if (req.originalUrl.indexOf('/api/v1/inventory/vehicle/') > - 1) {
+    const vehicle = await Vehicle.findById(req.params.vehicleId);
```

Comment:
Have we tested the case where the vehicles is not found? Does it send a 404 response or does it continue with an undefined vehicle variable and crash the middleware?

Reply:
That's a good catch. We handle errors like this vehicleController or saleController. However, this middleware gets called before. So i moved the errors check in the original controller files and added it in the middleware. This can be referenced in this [commit](https://github.com/purplesmurf1998/Autotracks/pull/64/commits/1b876ff643cf4ad0a6b8ae2a9f574bdfe7278863)